dask/array/chunk.py

```diff
@@ -396,3 +396,13 @@ def slice_with_int_dask_array_aggregate(idx, chunk_outputs, x_chunks, axis):
             idx_final if i == axis else slice(None) for i in range(chunk_outputs.ndim)
         )
     ]
+
+
+def getitem(obj, index):
+    result = obj[index]
+    try:
+        if obj.size >= 2 * result.size:
```

Comment:
I haven't really followed where this `getitem` is used, but is it possible that `result` is already a copy? For example, if `index` is a integer/boolean array, or anything triggering [advanced indexing](https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing), `result` would already be a copy, so taking another copy is unnecessary.

Reply:
Thanks for pointing this out @gjoseph92 . Yes, I noticed that when `idx = int` we don't have a memory shared. For example:
```
import operator
import numpy as np

>>>  arr =np.random.rand(1_000_000)
>>> test = operator.getitem(arr, 500)
>>> np.shares_memory(test, arr[500])
False

>>> test = operator.getitem(arr, slice(500, 501))
>>> np.shares_memory(test, arr[500:501])
True
```
 We should handle this and anything that triggers advancing indexing to avoid copying when it's not necessary.

Reply:
From reading the [`np.shares_memory` docs](https://numpy.org/doc/stable/reference/generated/numpy.shares_memory.html#numpy.shares_memory) it appears computing whether or not two arrays share memory can be expensive. Instead, I'll suggest we use the `owndata` flag on the sliced `result` array. This should be a decent proxy for if the array is a view of another array or not: 

```python
In [1]: import numpy as np

In [2]: x = np.arange(10)

In [3]: x.flags.owndata
Out[3]: True

In [4]: x[:2].flags.owndata
Out[4]: False

In [5]: x[:2].copy().flags.owndata
Out[5]: True
```